Allow passing mutiple formats to date-parser

smortex opened the pull request:
Thanks to this change, a single date-parser is able to process different date formats, making it easy to catch on when some programs change the way they log information.

Internal variables where renamed from `format` to `formats` to reflect the fact that the variable is now a list of formats, but the public template name has been kept back to avoid breaking existing configuration:

```
date-parser(format(
     "%FT%T.%f",
     "%F %T,%f",
     "%F %T"
    ));
```

_Suggested by @bazsi in https://github.com/balabit/syslog-ng/pull/2709#issuecomment-489568587_

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

Kokan commented:
@kira-syslogng ok to test

Kokan commented on modules/date/tests/test_date.c:
```diff
@@ -183,6 +183,46 @@ Test(date, test_date_with_additional_text_at_the_end)
+    { "2017-02-02 00:29:16",                0 },
+    { "2017-02-02 00:29:16,706",       706000 },
+    { "2019-05-04T21:55:46.989+02:00", 989000 },
+  };
```
Big thanks for the unit test!
Criterion has parametric test capabilities. A very simple example: https://github.com/balabit/syslog-ng/blob/master/tests/unit/test_findcrlf.c
Please instead of this loop either use that, or create the test one-by-one.

kira-syslogng commented:
Build SUCCESS

Kokan commented:
@smortex please fixup and I'll approve.

kira-syslogng commented:
Build FAILURE

bazsi commented:
This pull request **introduces 2 alerts** when merging 9fc4f123c2d6dfd7ebf0e2a0e04d86cb5b0061c7 into cf10b3c4a50d8feda50826b518428719fd974a1c - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-04253ee6fec040f923e6612e4fc3119f2e01b6e2)

**new alerts:**

* 2 for Ambiguously signed bit\-field member

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

Kokan commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build FAILURE

Kokan commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

Kokan commented:
@smortex the reports from LGTM is wrong, it reports issues from master :( but that is not blocking issue here.

bazsi commented on modules/date/date-grammar.ym:
```diff
@@ -70,23 +71,28 @@ date_parser_options
 date_parser_option
-        : KW_FORMAT '(' string ')'                  { date_parser_set_format(last_parser, $3); free($3); }
+        : KW_FORMAT '(' date_parser_formats ')'     { date_parser_set_format(last_parser, $3); }
         | KW_TIME_ZONE '(' string ')'               { date_parser_set_timezone(last_parser, $3); free($3); }
```
probably the setter name should be renamed to date_parser_set_formats()

bazsi commented on modules/date/date-parser.c:
```diff
@@ -85,16 +85,22 @@ _parse_timestamp_and_deduce_missing_parts(DateParser *self, WallClockTime *wct,
+  for (GList *item = self->date_formats; item; item = item->next)
+    {
+      remainder = wall_clock_time_strptime(wct, item->data, input);
 
```
I would probably keep this function just to parse a single timestamp and move the loop to a new function that iterates the list.

That would make a good place for the "date-parser message processing started" log message, which originally included the current format and right now only tells the user how many entries there are in the list. That is not very useful imho.

bazsi commented on modules/date/date-parser.c:
```diff
@@ -125,7 +131,7 @@ date_parser_process(LogParser *s,
             evt_tag_str ("input", input),
-            evt_tag_str ("format", self->date_format),
+            evt_tag_int ("formats", g_list_length(self->date_formats)),
             evt_tag_printf("msg", "%p", *pmsg));
```
this is which should include the actual format we are trying to parse against, with perhaps a bit of changes on the message text.

bazsi commented on modules/date/date-grammar.ym:
```diff
@@ -70,23 +71,28 @@ date_parser_options
+date_parser_formats
+        : string date_parser_formats                { $$ = g_list_prepend($2, $1); }
+        | string                                    { $$ = g_list_prepend(NULL, $1); }
+        ;
```
we do have a `string_list` rule IIRC, which could probably be used here.

bazsi replied:
I think it is ok to accept an empty list.

bazsi commented on modules/date/date-parser.c:
```diff
@@ -148,7 +154,7 @@ date_parser_clone(LogPipe *s)
   cloned = date_parser_new(log_pipe_get_config(&self->super.super));
-  date_parser_set_format(cloned, self->date_format);
+  date_parser_set_format(cloned, g_list_copy_deep(self->date_formats, (GCopyFunc) g_strdup, NULL));
   date_parser_set_timezone(cloned, self->date_tz);
```
hopefully there's a string list helper that clones the list, if not one could be created.

bazsi commented on modules/date/date-parser.c:
```diff
@@ -161,7 +167,7 @@ date_parser_free(LogPipe *s)
 
-  g_free(self->date_format);
+  g_list_free_full(self->date_formats, g_free);
   g_free(self->date_tz);
```
again, a good candidate for extraction into a the string-list set of functions.

bazsi replied:
This could be converted to string_list_free()

smortex replied:
Ooops, missed that one!

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

smortex commented:
@bazsi i tried to update the cisco parser but am not sure that my changes are properly working: I used the examples in the comments as log messages and observe similar results but not what I expected, so I am not confident that the behavior is correct.  Is the kira failure related?

You should be allowed to add commits to my branch, feel free to do so if you spot something wrong :wink:

bazsi commented:
I was kind of hoping we could get rid of the regexps by using alternate
formats, I see you didn't, is there anything missing?

The new light test framework covers the cisco-parser() so if you execute
make pytest-check it should test with the patterns in the scl file.



Romain Tartière <notifications@github.com> ezt írta (időpont: 2019. jún.
15., Szo 2:51):

> @bazsi <https://github.com/bazsi> i tried to update the cisco parser but
> am not sure that my changes are properly working: I used the examples in
> the comments as log messages and observe similar results but not what I
> expected, so I am not confident that the behavior is correct. Is the kira
> failure related?
>
> You should be allowed to add commits to my branch, feel free to do so if
> you spot something wrong 😉
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2779?email_source=notifications&email_token=AAFOK5T72IGZK52XMILL3QLP2Q4J5A5CNFSM4HXVAMQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXYM3MI#issuecomment-502320561>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFOK5VDWCAOBJW3UNZGUUDP2Q4J5ANCNFSM4HXVAMQQ>
> .
>

smortex commented:
> I was kind of hoping we could get rid of the regexps by using alternate formats, I see you didn't, is there anything missing? The new light test framework covers the cisco-parser() so if you execute make pytest-check it should test with the patterns in the scl file.

Meh, the test suite use the *installed version* of the scl files.  When I understood that, I could adjust the configuration file and test the changes.  With minimal modifications I could fix the code in the way you expected @bazsi \o/

kira-syslogng commented:
Build FAILURE

bazsi commented on scl/cisco/plugin.conf:
```diff
@@ -37,26 +37,11 @@
-            parser { date-parser(format('%b %d %Y %H:%M:%S') template("$1")); };
-        };
+        parser { date-parser(format(' *%b %d %H:%M:%S %p.%f', ' *%b %d %H:%M:%S.%f', ' *%b %d %Y %H:%M:%S.%f') template(`template`)); };
     };
```
Much much better! thank you. btw, we could theoretically add all of these formats with a %Z that should parse the timezone as well. something like:

```
parser { date-parser(format(' *%b %d %H:%M:%S %p.%f %Z', ' *%b %d %H:%M:%S %p.%f', ' *%b %d %H:%M:%S.%f %Z', ' *%b %d %H:%M:%S.%f', ' *%b %d %Y %H:%M:%S.%f %Z', ' *%b %d %Y %H:%M:%S.%f') template(`template`)); };
```

Although %Z in our current strptime() implementation is pretty limited (it only recognizes UTC, GMT and the name of the local timezone), that can be extended later, still I think it is better to add it here, and potentially improve strptime() at a later point.

Consider this optional, I already think this is a huge improvement.

bazsi replied:
btw , why do you have a space in front? we didn't have that in the previous implementation.

smortex replied:
For the extra space, this was probably an error on my side when juggling with this branch and the one where I was trying to fix the test suite on my test machine running FreeBSD.

Adding the time zone, I spot another issue related to 12-hour clock formats and hopefully fixed it (H -> %I).  BTW, do you have some cisco logs with time zone information to share to extend the test suite?

kira-syslogng commented:
Build FAILURE

Kokan commented:
@smortex this is an actual cisco parser test failing, tomorrow I'll check why and share details.

Kokan commented on scl/cisco/plugin.conf:
```diff
@@ -37,25 +37,17 @@
-        if {
-            # timestamp from Cisco Unified Call Manager, example "Jun 14 11:57:27 PM.685 UTC"
-            # NOTE: drops msecs and timezone
-            filter { match('^[.*]?([A-Za-z]{3} [0-9 ]\d \d{2}:\d{2}:\d{2} ((AM)|(PM)))' template(`template`) flags(store-matches)); };
```
These filters are not *read-only* from the message perspective, because of the `store-matches` instruments the filter to save the matched content. Like the `(...)` blocks, and place them into the `$1, $2, $3, ..., $n`.

One way to resolve it is to keep one filter that accepts the same timestamps as either of those three; and also stores the matched timestamp part into `$1`.

Btw the sets of log message it is tested against. (drops a few of these)
```
<189>91: *Aug 27 14:22:52.787: %SYS-5-CONFIG_I: Configured from console by console
<190>61: *Aug 27 11:21:30.639: %SYS-6-LOGGINGHOST_STARTSTOP: Logging to host 10.10.31.78 started - CLI initiated
<189>90: *Aug 27 14:22:09: %SYS-5-CONFIG_I: Configured from console by console
<189>92: *Aug 27 14:23:23 UTC: %SYS-5-CONFIG_I: Configured from console by console
<189>93: *Aug 27 14:23:40.439 UTC: %SYS-5-CONFIG_I: Configured from console by console
<189>97: *Aug 28 2015 02:25:01.199 Buthan: %SYS-5-CONFIG_I: Configured from console by console
<189>98: %SYS-5-CONFIG_I: Configured from console by console
<189>32: 0.0.0.0: *Apr 29 13:59:12.491: %SYS-5-CONFIG_I: Configured from console by console
<189>102: R1: *Aug 28 02:27:33.231 Buthan: %SYS-5-CONFIG_I: Configured from console by console
```

smortex replied:
Thanks!  So the milliseconds seems to be optional, and "Buthan" as the timezone is rejected as said by @bazsi here:
https://github.com/balabit/syslog-ng/pull/2779#discussion_r294081186

I guess I will have to drop the timezone support for now.

bazsi commented:
Yeah, but the only reason we extract $1 etc is to pass it to date-parser()
so if the date-parser can do it on its own, thats an improvement.

Kókai Péter <notifications@github.com> ezt írta (időpont: 2019. jún. 17., H
11:49):

> *@Kokan* commented on this pull request.
> ------------------------------
>
> In scl/cisco/plugin.conf
> <https://github.com/balabit/syslog-ng/pull/2779#discussion_r294218133>:
>
> >  #
>  block parser cisco-timestamp-parser(template()) {
>      channel {
> -        if {
> -            # timestamp from Cisco Unified Call Manager, example "Jun 14 11:57:27 PM.685 UTC"
> -            # NOTE: drops msecs and timezone
> -            filter { match('^[.*]?([A-Za-z]{3} [0-9 ]\d \d{2}:\d{2}:\d{2} ((AM)|(PM)))' template(`template`) flags(store-matches)); };
>
> These filters are not *read-only* from the message perspective, because
> of the store-matches instruments the filter to save the matched content.
> Like the (...) blocks, and place them into the $1, $2, $3, ..., $n.
>
> One way to resolve it is to keep one filter that accepts the same
> timestamps as either of those three; and also stores the matched timestamp
> part into $1.
>
> Btw the sets of log message it is tested against. (drops a few of these)
>
> <189>91: *Aug 27 14:22:52.787: %SYS-5-CONFIG_I: Configured from console by console
> <190>61: *Aug 27 11:21:30.639: %SYS-6-LOGGINGHOST_STARTSTOP: Logging to host 10.10.31.78 started - CLI initiated
> <189>90: *Aug 27 14:22:09: %SYS-5-CONFIG_I: Configured from console by console
> <189>92: *Aug 27 14:23:23 UTC: %SYS-5-CONFIG_I: Configured from console by console
> <189>93: *Aug 27 14:23:40.439 UTC: %SYS-5-CONFIG_I: Configured from console by console
> <189>97: *Aug 28 2015 02:25:01.199 Buthan: %SYS-5-CONFIG_I: Configured from console by console
> <189>98: %SYS-5-CONFIG_I: Configured from console by console
> <189>32: 0.0.0.0: *Apr 29 13:59:12.491: %SYS-5-CONFIG_I: Configured from console by console
> <189>102: R1: *Aug 28 02:27:33.231 Buthan: %SYS-5-CONFIG_I: Configured from console by console
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2779?email_source=notifications&email_token=AAFOK5V3CRUMXRBF3T5YTYTP25M2VA5CNFSM4HXVAMQ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3WLURY#pullrequestreview-250395207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFOK5R2VWJZ4HCYCOXC4Z3P25M2VANCNFSM4HXVAMQQ>
> .
>

kira-syslogng commented:
Build SUCCESS

smortex commented:
@Kokan, thanks to the provided sample log files, I could fix the support for timestamps without milliseconds.  I also chose to revert back to ignoring the time zone in the timestamp (as was done before) because the current support is not compatible with the time zone present in the samples you provided and would cause unacceptable backward incompatibility without additional work in the time parser.

I did the tests by adding your provided samples to my configuration, but I have not committed it in the repo.  Please tell me if you would like me to do so.

Last, there are a lot of fixups and reverted changes in the last ~10 commits of the PR.  Please tell me when you prefer me to rework/squash them: I feel like it's becoming harder to follow the current progress then it would if the commits where "clean" :wink:

bazsi commented:
Please merge the fixups where appropriate. We can work on the timezone support in a separate PR.

I don't understand this in your comment: "I did the tests by adding your provided samples to my configuration, but I have not committed it in the repo. Please tell me if you would like me to do so."

smortex commented:
Hi @bazsi,

I squashed the commits, reworded them to follow conventions and rebased them on top of master.

---------

> I don't understand this in your comment: "I did the tests by adding your provided samples to my configuration, but I have not committed it in the repo. Please tell me if you would like me to do so."

I just mean that I did not added the provided examples to `tests/python_functional/functional_tests/parsers/app-parser/test_topic_syslog_raw.py` because I understood they where part of another [private] test suite run by jenkins and that would be redundant.  If I misunderstood and you expected me to add these tests to the `test_topic_syslog_raw.py` file, please tell me so and I will do it :smile:

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

Kokan commented:
@smortex It is not required to add those sample/test cases.

Kokan commented:
@kira-syslogng retest this please

bazsi commented:
Those samples are part of the private performance test suite of the syslog-ng team, so it's their decision. 

On the other hand the sample with "Buthan" in the timezone field, it's definitely wrong. Buthan is spelled Bhutan and it is not an official timezone name. It is called "Bhutan Time", abbreviated as BTT

@pzoleex do you know where that sample is coming from? As far as I understand Cisco routers allow any user specified string to be set for the timezone field.

kira-syslogng commented:
Build SUCCESS
